crates/oxc_semantic/src/node/mod.rs

```diff
@@ -50,11 +51,16 @@ impl<'a> SemanticNode<'a> {
     #[must_use]
     pub fn strict_mode(&self, scope: &Scope) -> bool {
         // All parts of a ClassDeclaration or a ClassExpression are strict mode code.
-        scope.strict_mode() || self.in_class()
+        scope.strict_mode() || (self.in_class() && !self.in_decorator())
```

Comment:
I'm not getting this 😕 Can you give me an example for which this applies?

Reply:
Let's add an explanation comment for this, and a reference to the proposal or the spec.

Reply:
Without this we couldn't parse https://github.com/tc39/test262/blob/main/test/language/expressions/class/decorator/syntax/valid/decorator-parenthesized-expr-identifier-reference-yield.js
<img width="831" alt="Screen Shot 2023-03-15 at 11 37 06 PM" src="https://user-images.githubusercontent.com/42916814/225515681-30efab68-d099-47fd-981d-1273a85a8168.png">

Reply:
https://github.com/Boshen/oxc/blob/164c2831e10e15c4288b210f12f0eda812bfdd74/crates/oxc_ast/src/visit.rs#L358-L363

I think it's because of the visiting logic, decorators are not part of the class?

Reply:
Two thoughts: 
* do we need to visit decorators?
* given decorators are transpiled, then the visiting order should be as if we are visiting the transpiled code

Reply:
> * do we need to visit decorators?

I think so. This code should produce error in the semantic analysis pass:

```javascript
"use strict"

@yield class A{}
```

But I'd need to take another look. The current approach would accept 
```javascript
class A {
@yield a() {}
}
```
but babel rejects it. Seems only the class decorator are outside of the class.

> * given decorators are transpiled, then the visiting order should be as if we are visiting the transpiled code

How would we do that given the AST structure? I don't know if the visiting order matters for the purpose of semantic analysis. Seems that decorators are simply an expression referring to a function (or partial application of a function?).

Reply:
Let's move the order of the visit, visit decorator then enter class. Given the fact it transpiles down to a class wrapped by a function.